imap/jmap_ical.c

```diff
@@ -4106,29 +4106,128 @@ calendarevent_from_ical(icalcomponent *comp,
+            // made a copy of it, yet. Clone it now.
+            myical = icalcomponent_clone(original_ical);
+        }
+
+        icalcomponent *vevent;
+        for (vevent = icalcomponent_get_first_component(myical,
+                    ICAL_VEVENT_COMPONENT);
+             vevent;
+             vevent = icalcomponent_get_next_component(myical,
+                    ICAL_VEVENT_COMPONENT)) {
+
+            // Make sure that the mandatory UID property is set.
+            if (!icalcomponent_get_uid(vevent)) {
+                if (pass == 2) {
+                    // Generate UID property by hashing normalized component
+                    icalcomponent *norm_comp = icalcomponent_clone(vevent);
+                    icalcomponent_normalize_x(norm_comp);
+                    char hexbuf[JMAPICAL_SHA1HEXSTR_LEN];
+                    sha1hexstr(icalcomponent_as_ical_string(norm_comp), hexbuf);
+                    buf_setcstr(&buf, "nouid");
+                    buf_appendcstr(&buf, hexbuf);
+                    icalcomponent_add_property(
+                        vevent, icalproperty_new_uid(buf_cstring(&buf)));
+                    icalcomponent_free(norm_comp);
+                }
+                else needs_rewrite = 1;
+            }
+
+            // Remove METHOD:PUBLISH if no ORGANIZER is set
+            if (method == ICAL_METHOD_PUBLISH &&
+                !icalcomponent_get_first_property(vevent,
+                                                  ICAL_ORGANIZER_PROPERTY)) {
+                if (pass == 2) {
+                    // XXX this resets the property iterator of ical
+                    icalproperty *prop = icalcomponent_get_first_property(myical,
+                            ICAL_METHOD_PROPERTY);
+                    if (prop) {
+                        icalcomponent_remove_property(myical, prop);
+                        icalproperty_free(prop);
+                    }
+                }
+                else needs_rewrite = 1;
+            }
+        }
+
+        if (!needs_rewrite)
+            break;
+    }
+
+    buf_free(&buf);
+    *icalp = myical;
+}
+
 EXPORTED json_t*
 jmapical_tojmap_all(icalcomponent *ical, hash_table *props,
                     struct jmapical_ctx *jmapctx)
 {
+    json_t *jsevents = json_array();
+    icalcomponent *myical = ical;
     icalcomponent *comp;
-    size_t ncomps = 0;
+    struct buf buf = BUF_INITIALIZER;
+
+    if (jmapctx && jmapctx->from_ical.repair_broken_ical) {
+        repair_broken_ical(&myical);
+    }
 
     // Count the total number of components
-    for (comp = icalcomponent_get_first_component(ical, ICAL_VEVENT_COMPONENT);
+    size_t ncomps = 0;
+    for (comp = icalcomponent_get_first_component(myical, ICAL_VEVENT_COMPONENT);
          comp;
-         comp = icalcomponent_get_next_component(ical, ICAL_VEVENT_COMPONENT)) {
+         comp = icalcomponent_get_next_component(myical, ICAL_VEVENT_COMPONENT)) {
              ncomps++;
```

Comment:
Unrelated to the PR, but this for{} loop can be replaced with icalcomponent_count_components()

Reply:
Thanks, I was not aware of this function! I updated the code accordingly.